crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,197 @@
+    (and excludes all other functions, even if they have the same signature).
+- A "literal module type": a type representing a single known module.
+
+Since it is known that all inhabitants of a given singleton type share the same memory address,
+singleton types in unions can be safely narrowed by identity,
+using [the operators `is` and `is not`](https://snarky.ca/unravelling-is-and-is-not/):
+
+```py
+def f(x: str | None):
+    if x is None:
+        ...  # x can be narrowed to `str`
+    else:
+        ...  # x can be narrowed to `None`
+```
+
+All singleton types are also sealed types, closed types and single-value types
+(see below for definitions).
+
+## Sealed types
+
+A sealed type is a type for which it is known
+that there is only a finite and pre-determined set of inhabitants at runtime.
+It follows from this that for every sealed type,
+there is also only a finite and pre-determined set of *subtypes* of that type.
+
+A sealed type in Python has the property
+that its proper subtypes[^2] are all disjunct from one another,
+and the union of all the proper subtypes of a sealed type is exactly equal to the sealed type.
+
+All singleton types are sealed types; however, not all sealed types are singleton types.
+
+Examples of sealed types (other than the singleton types listed above) are:
+
+- `bool`:
+
+    - The only inhabitants of `bool` at runtime are the constants `True` and `False`.
+    - The only proper subtypes of `bool` are `Literal[True]` and `Literal[False]`.
+        `Literal[True]` and `Literal[False]` are disjunct types, and their union is
+        exactly equal to `bool`.
+
+- Enums: consider the following enum class:
+
+    ```py
+    from enum import Enum, auto
+
+    class Foo(Enum):
+        X = auto()
+        Y = auto()
+    ```
+
+    - The only inhabitants of the `Foo` type at runtime are the enum `Foo`'s members:
+        `Foo.X` and `Foo.Y`
+    - The only proper subtypes of `Foo` are `Literal[Foo.X]` and `Literal[Foo.Y]`.
+        `Literal[Foo.X]` and `Literal[Foo.Y]` are disjunct, and the union `Literal[Foo.X, Foo.Y]`
+        is exactly equal to the type `Foo`.
+
+Because a singleton type is equivalent to the union of all of its proper subtypes,
+for any given singleton type `X` where the only proper subtypes of `X` are `A`, `B` and `C`,
+`X & ~A == B | C`. To give a Python example:
+
+```py
+from enum import Enum, auto
+
+class X(Enum):
+    A = auto()
+    B = auto()
+    C = auto()
+
+
+def f(var: X):
+    ...  # the type of `var` here is `X` (which is equivalent to `Literal[X.A, X.B, X.C]`)
+
+    if var is X.A:
+        ...  # var is narrowed to `Literal[X.A]`
+    else:
+        ...  # var is narrowed to `Literal[X.B, X.C]`
+```
+
+## Single-value types
+
+A single-value type is a type for which it is known
+that all inhabitants of the type are equivalent with respect to their underlying value.
+All singleton types are single-value types, but not all single-value types are singleton types.
```

Comment:
One thing that could be clarified here: Is `Never` a single-value type or not? The way I implemented `is_single_valued()`, `Never` is excluded. So that definition would have to include something like "A single-value type is a non-empty type for which …"